src/main/java/com/coherentsolutions/training/aqa/java/web/nikonova/browsers/BrowsersSettings.java

```diff
@@ -0,0 +1,103 @@
+                firefoxOptions = new FirefoxOptions();
+                firefoxOptions.setPlatformName(BrowsersProperties.getBrowserProperties(String.format("%s.firefox.browser.platform", environment)));
+                firefoxOptions.setBrowserVersion(BrowsersProperties.getBrowserProperties(String.format("%s.firefox.browser.version", environment)));
+                driver.set(ThreadGuard.protect(new RemoteWebDriver(new URL(BrowsersProperties.getBrowserProperties("url.grid")), firefoxOptions)));
+            } else {
+                throw new Exception("Browser is unavailable");
+            }
+        } else if (environment.equals(BrowsersProperties.getBrowserProperties("env.sauceLabs"))) {
+            if (browser.equals(BrowsersProperties.getBrowserProperties("browser.chrome"))) {
+                driver.set(ThreadGuard.protect(sauceLabsTestOnChrome()));
+            } else if (browser.equals(BrowsersProperties.getBrowserProperties("browser.firefox"))) {
+                driver.set(ThreadGuard.protect(sauceLabsTestOnFirefox()));
+            } else {
+                throw new Exception("Browser is unavailable");
+            }
+        } else {
+            throw new Exception("Environment is unavailable");
+        }
+    }
+
+    private RemoteWebDriver sauceLabsTestOnFirefox() throws IOException {
+        String username = System.getenv("SAUCE_USERNAME");
+        String accessKey = System.getenv("SAUCE_ACCESS_KEY");
+        MutableCapabilities sauceOptions = new MutableCapabilities();
+
+        sauceOptions.setCapability("username", username);
+        sauceOptions.setCapability("accessKey", accessKey);
+
+        firefoxOptions = new FirefoxOptions();
+        firefoxOptions.setPlatformName(BrowsersProperties.getBrowserProperties(String.format("%s.firefox.browser.platform", environment)));
+        firefoxOptions.setBrowserVersion(BrowsersProperties.getBrowserProperties(String.format("%s.firefox.browser.version", environment)));
+        return new RemoteWebDriver(new URL(BrowsersProperties.getBrowserProperties("url.sauceLabs")), firefoxOptions);
+    }
+
+    private RemoteWebDriver sauceLabsTestOnChrome() throws IOException {
+        String username = System.getenv("SAUCE_USERNAME");
+        String accessKey = System.getenv("SAUCE_ACCESS_KEY");
+        MutableCapabilities sauceOptions = new MutableCapabilities();
+
+        sauceOptions.setCapability("username", username);
+        sauceOptions.setCapability("accessKey", accessKey);
+
+        chromeOptions = new ChromeOptions();
+        chromeOptions.setPlatformName(BrowsersProperties.getBrowserProperties(String.format("%s.chrome.browser.platform", environment)));
+        chromeOptions.setBrowserVersion(BrowsersProperties.getBrowserProperties(String.format("%s.chrome.browser.version", environment)));
+        return new RemoteWebDriver(new URL(BrowsersProperties.getBrowserProperties("url.sauceLabs")), chromeOptions);
+    }
```

Comment:
Please take a look on these methods. The first five lines are exactly the same, in order not to copy-paste, it’s easier to create a separate method and reuse the code